### ISSUE TITLE
[Case Analysis] Detect the outliers in data and the weakness in algorithm

### ISSUE THREAD

**Issue by L-M-Sherlock** [open-spaced-repetition/fsrs4anki#275]
Here are 2 decks that seem to suffer from outlier detection. These two have shown negative results both when using your version without any changes and when combining theo utlierdetection with 4.0 Alpha.
[For the optimizer.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/11598656/For.the.optimizer.zip)

_Originally posted by @Expertium in https://github.com/open-spaced-repetition/fsrs4anki/issues/239#issuecomment-1568151387_

**Comment by Expertium** [comment]
First I ran the original version (I copied the code to a colab file and saved it on my Google Drive because your colab link doesn't work). Then I added the outlier detection to 4.0 Alpha. Here are my results (I always use n_splits=5, n_epoch=5 and batch_size=512)
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/778a7dac-14ac-4beb-b400-aa1585b0eaf4)
Huh, strange. I re-ran my copy again and got a somewhat different result.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/71ec0192-2b0c-41ca-8b90-e0f0e9b2ba50)

**Comment by L-M-Sherlock** [comment]
> Huh, strange. I re-ran my copy again and got a somewhat different result.

Sorry for that. I forget to push my code to the remote before I share you with the link.

**Comment by L-M-Sherlock** [comment]
Oh, the link is still incorrect. Wait for a minutes

**Comment by L-M-Sherlock** [comment]
https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/mini-batch_initial-stability.ipynb

**Comment by Expertium** [comment]
> https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/mini-batch_initial-stability.ipynb

I got the following result (n_splits=5, n_epoch=5, batch_size=512):
Logloss=0.3542 (better than 4.0 Alpha)
RMSE=0.0651 (worse than 4.0 Alpha)

**Comment by L-M-Sherlock** [comment]
> I got the following result for my deck "Hard English" (n_splits=5, n_epoch=5, batch_size=512):
> Logloss=0.3542 (better than 4.0 Alpha)
> RMSE=0.0651 (worse than 4.0 Alpha)

Oh, it is not a fair comparison. We should compare it with mini-batch baseline. Or we should compare the 4.0 Alpha with 4.0 Alpha with filter.

By the way, I find that increase the n_splits will worsen the performance of initial stability. Because some splits don't include some first rating data.

**Comment by Expertium** [comment]
> Oh, it is not a fair comparison. We should compare it with mini-batch baseline. Or we should compare the 4.0 Alpha with 4.0 Alpha with filter.

Then I suggest adding the outlier detection part of the code to 4.0 Alpha and making a new colab link

**Comment by L-M-Sherlock** [comment]
https://github.com/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/fsrs4anki_optimizer_alpha_filter.ipynb
https://github.com/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/fsrs4anki_optimizer_alpha.ipynb

**Comment by Expertium** [comment]
Why is this part of the code different?
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/7d540c44-bdb2-4e23-8636-d4488339e9dd)

**Comment by L-M-Sherlock** [comment]
> Why is this part of the code different?

Please see the comment at https://github.com/open-spaced-repetition/fsrs4anki/issues/275#issuecomment-1568277930

**Comment by Expertium** [comment]
fsrs4anki_optimizer_alpha_filter.ipynb, deck "Hard Egnlish"
logloss=0.3344
RMSE=0.0553
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/99237686-d46d-4b7c-85ef-1c502e886f83)

fsrs4anki_optimizer_alpha.ipynb
logloss=0.3623
RMSE=0.0524
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/44e1faec-689a-4542-b140-d0c2498c1672)

On a qualitative level, the blue calibration line looks more random when filters are used.

**Comment by L-M-Sherlock** [comment]
> fsrs4anki_optimizer_alpha_filter.ipynb, deck "Hard Egnlish"
> logloss=0.3344
> RMSE=0.0553
> ![image](https://user-images.githubusercontent.com/83031600/241990984-99237686-d46d-4b7c-85ef-1c502e886f83.png)

I don't know why your result is different from mine:

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/b595ba8f-dfb9-4567-bbf2-738a90c7d203)

I also use `Main__English__Hard English_27.04.2023.apkg` to run the optimizer.

Oh, I forget the `timezone` and `next_day_starts_at`.

**Comment by L-M-Sherlock** [comment]
> I really want to see how [this](https://github.com/open-spaced-repetition/fsrs4anki/issues/239#issuecomment-1568266213) will perform, so I would be thankful if you implemented it instead of the interquartile range.

OK, I will consider it later.

**Comment by L-M-Sherlock** [comment]
@Expertium, could you provide `timezone` and `next_day_starts_at`?

**Comment by Expertium** [comment]
timezone = 'Europe/Moscow'

next_day_starts_at = 5

**Comment by L-M-Sherlock** [comment]
> I really want to see how [this](https://github.com/open-spaced-repetition/fsrs4anki/issues/239#issuecomment-1568266213) will perform, so I would be thankful if you implemented it instead of the interquartile range.

I have implemented here: https://github.com/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/mini-batch_initial-stability_LOF.ipynb

And I update the code of initial stability here: https://github.com/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/mini-batch_initial-stability.ipynb

Now they have the best performance:

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/e56e502b-dfd3-4629-b09f-4192c8142c41)

**Comment by Expertium** [comment]
"Ultimate Geography", fsrs4anki_optimizer_alpha.ipynb (n_splits=5, n_epoch=5, batch_size=512)
logloss=0.5166
RMSE=0.0593
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/d647164d-c2ab-41f7-a934-d8b5aa65e587)

"Ultimate Geography", mini-batch_initial-stability_LOF.ipynb
logloss=0.5512
RMSE=0.1226
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/2f92422b-4901-4f37-b524-bc375eb737a9)

This is strange. This one deck doesn't benefit from outlier removal at all. It's possible that we will have to implement "double optimization", with and without outlier detection, and then select the best set of parameters based on RMSE. But only if we don't find a way to make outlier removal consistent in how it affects RMSE.

**Comment by L-M-Sherlock** [comment]
There seemly are many data in the low retention area. I will check them.

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/fbc1f66f-418c-40cf-a6c5-461f98bdacd2)

**Comment by L-M-Sherlock** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/73a1f497-0bd1-456a-aa83-1dfabab12bc8)

Solved. Just increase the learning rate to `1e-1`.

**Comment by L-M-Sherlock** [comment]
> Solved. Just increase the learning rate to `1e-1`.

I locate the problem. It is due to the pre-train stage, which doesn't fit the initial stability enough. I fix it here:

https://github.com/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/mini-batch_initial-stability_LOF.ipynb

**Comment by Expertium** [comment]
> https://colab.research.google.com/github/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/mini-batch_initial-stability_LOF.ipynb

I tested it with lr=1e-1 (which isn't very fair since in 4.0 Alpha it's 4e-2). It still performs mildly worse than 4.0 Alpha for Ultimate Geography, though better than before.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/438f16bf-03fa-46fa-91be-0869fec5ee57)

I tested it on my decks and collection + your collection + user1823's collection, and finally obtained statistically significant results. It makes results worse for your collection, and for user1823 it makes resutls negligibly better.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/f2ac0444-692a-4a28-9c72-2ffc4f59c044)

I suggest implementing these modifications to further improve the classification of outliers: https://github.com/open-spaced-repetition/fsrs4anki/issues/275#issuecomment-1568795623.
Also, is pre-training necessary? I thought you had ditched that idea.

**Comment by L-M-Sherlock** [comment]
> Also, is pre-training necessary? I thought you had ditched that idea.

I get it back due to the initial stability idea.

For the removing outliers, I think we should pay more attention to the causes of them. Some of them are overdue, but others are applied with different scheduling settings of the default algorithm. Such as gradual interval and leaning steps.

**Comment by Expertium** [comment]
> For the removing outliers, I think we should pay more attention to the causes of them. Some of them are overdue, but others are applied with different scheduling settings of the default algorithm. Such as gradual interval and leaning steps.

Yes, but I still think that we should use LOF to its full potential.

**Comment by Expertium** [comment]
@L-M-Sherlock for some reason, the first interval for "Again" is longer than for "Hard" and "Good"
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/e28cfd3f-f952-4cee-b9bc-4a4d69d75ce6)
And D doesn't change after the first review I tried this:
`new_d = self.mean_reversion(state[:,1], new_d)`
Basically, instead of using the initial value of D in the mean reversion, we use the previous value. It improved RMSE, but the weird behavior of D and intervals remains.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/9cfd15e7-eeec-4966-a375-8b3ff8a148fd)

**Comment by Expertium** [comment]
I've tried different clamping for self.w[5] and that also doesn't solve the problem.

**Comment by Expertium** [comment]
Ok, I figured out how to enforce that.
```
            w[13] = w[13].clamp(0.01, 20)
            w[14] = w[14].clamp(w[13], 20)
            w[15] = w[15].clamp(w[14], 20)
            w[16] = w[16].clamp(w[15], 20)
```
But it only made results worse.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/0f96283c-d9aa-4a0c-8502-468ebd17b01a)

**Comment by L-M-Sherlock** [comment]
It is due to the lack of some first ratings. I find that you never used `again` in the first learning in the deck.

**Comment by Expertium** [comment]
> I find that you never used `again` in the first learning in the deck.

Yeah, that's my habit.

**Comment by user1823** [comment]
Lately, a huge emphasis has been placed on filtering out the outliers. So, I would like to share my views on it.

Broadly speaking, it is a good idea to remove the outliers before training, but we must remember that many outliers are caused by the limitations of the algorithm. So, our aim should be to identify such outliers and then try to improve the algorithm to fit them (instead of just trying to filter them out).

We must keep in mind that filtering out happens in the optimizer but not in the scheduler/helper. This means that the filtered cards continue to be scheduled incorrectly. This means that filtering the outliers is **hiding the problem rather than solving it**. However, I do believe that there is still some benefit of filtering as it prevents those cards from affecting the performance of the algorithm for the other cards.

But, if we improve the algorithm to deal with these outlier cards, those cards will also start being scheduled in the expected manner.

Sherlock said the following, which aligns with my thoughts:
> For the removing outliers, I think that we should pay more attention to the causes of them. Some of them are overdue, but others are applied with different scheduling settings of the default algorithm. Such as gradual interval and leaning steps.

Expertium shared the following idea. I think that it should be implemented so that our comparisons are fairer. In other words, I think that we should compare the loss and RMSE for the complete collection and not that for the filtered collection.

> Even if after removing outliers the algorithm performs better on a dataset without outliers, can we really be sure that it performs better on the same dataset with outliers?
So, suppose there are two datasets, A and B. A has no outliers, and B has outliers. Alg(A) was trained on A, Alg(B) was trained on B. We know that Alg(A) performs well on A, but how can we be sure that it will also perform well on B?
In order to test that, we need to calculate the log-loss and RMSE both for filtered and unfiltered datasets. So, I think that we need another section like 4.1 (call it 4.1.5, for example) and another section like 4.2, so we can see how an algorithm that was trained on a dataset with no outliers performs on the original dataset in terms of log-loss and RMSE.

**Comment by Expertium** [comment]
I think this issue can be closed now
